### ISSUE TITLE
Arrow keys changing screen orientation on fullscreen

### ISSUE THREAD

**Issue by Jeslandor** [moonlight-stream/moonlight-qt#505]
**Describe the bug**
While on fullscreen with the special shortkeys capture option enabled the cursor keys started changing my client pc screen rotation instead of acting as normal arrow keys.

Client: Moonlight portable on Windows 10 x64
Host: Latest version of Geforce Experience on Windows 7 x64

Sadly I don't know how to reproduce this as for now. I actually only wanted to bring attention that this happened. Just that. If I get the same issue again and get more information I will update.

On a different side note I want to notice too that special shortkeys new feature seem to work well on a Windows 7 host, included "Alt-Tab". So I assume the issue described in the changelog refers only when used Windows 7 as client.

Thank you very much and sorry for any inconvenience.

**Comment by cgutman** [comment]
If you're using an Intel iGPU, perhaps you triggered this hotkey - https://www.intel.com/content/www/us/en/support/articles/000005491/graphics.html

**Comment by Jeslandor** [comment]
> If you're using an Intel iGPU, perhaps you triggered this hotkey - https://www.intel.com/content/www/us/en/support/articles/000005491/graphics.html

Yeah, it certainly is partially related to this you mention.

For what I have tested since I opened the issue I have found that it happens when changing from fullscreen to window mode and back on realtime using the hotkey for that.

Once you are in fullscreen mode again, capture of special hotkeys stops working and things like Win, Alt-Tab, etc are reverted back to the PC running the Moonlight client. 

And this is when the issue is triggered. It seems that the client PC still thinks Ctrl and Alt are still pressed (maybe Shift too, don't know) so when you try to use the arrows the intel graphics screen rotation hotkey of the client PC is triggered.

I guess making sure a release key command is send on those would solve the most problematic part of the issue easily. Although don't know about the special host system hotkeys capture.

For now the only way to work again after that without the issue (and also with special windows hotkeys working again) is to disconnect session and resume it again.

Anyway, I also want to thank you very much for the reply, the useful help and of course for this incredible piece of software that Moonlight is. :)

**Comment by cgutman** [comment]
> For what I have tested since I opened the issue I have found that it happens when changing from fullscreen to window mode and back on realtime using the hotkey for that.
> 
> Once you are in fullscreen mode again, capture of special hotkeys stops working and things like Win, Alt-Tab, etc are reverted back to the PC running the Moonlight client.
> 
> And this is when the issue is triggered. It seems that the client PC still thinks Ctrl and Alt are still pressed (maybe Shift too, don't know) so when you try to use the arrows the intel graphics screen rotation hotkey of the client PC is triggered.
> 

I gave this a try on my Windows 10 PC and didn't encounter that issue. Switching from fullscreen to windowed to fullscreen didn't result in the loss of special hotkey capture, or result in any keys stuck down on the client.

> I guess making sure a release key command is send on those would solve the most problematic part of the issue easily. Although don't know about the special host system hotkeys capture.
> 
> For now the only way to work again after that without the issue (and also with special windows hotkeys working again) is to disconnect session and resume it again.
> 

Can you post a screenshot of your Moonlight settings so I can be sure I'm reproducing the exact scenario you are testing?

Does the bug happen both in fullscreen and borderless windowed mode?

Please also post a log from Moonlight with SDL events by opening Command Prompt and running:
```
set SDL_EVENT_LOGGING=1
"C:\Program Files\Moonlight Game Streaming\Moonlight.exe"
```
Go ahead and trigger the bug, then quit Moonlight. The log will be in your `%TEMP%` folder. 
NOTE: The log will include key events, so don't type anything sensitive into the stream while running this test.

**Comment by cgutman** [comment]
I think this is fixed now that I've upstreamed the keyboard grab code to SDL. There were some edge cases that weren't really possible to handle before.

Please try a CI build and see if the problem is gone: https://ci.appveyor.com/project/cgutman/moonlight-qt/builds/37624044/job/df9liuo5bl15ntpn/artifacts

**Comment by Jeslandor** [comment]
Damn, the issue is closed automatically now.

First of all I really want to apologize. It was not fair on my side. I was distracted by life stuff and completely forgot about this as I didn't use again moonlight since last time I wrote here.

Also thank you very much for all the effort I see you have put in the issue and replies. Again, I really really sorry.

I am not sure now if I should write here anyway or create a new issue. I guess I will continue here and wait to see if it's noticed and if not I will create then a new issue.

Ok, regarding the issue, I downloaded the x64 version linked from the last reply. Sadly what I have found using that version is special hotkeys are not working anymore in the host PC independently if the option is selected or not. It behaves exactly as previous to 3.0.0 version. Pressing Windows key or any other special combination only affects windows on the client PC.

What I have found when going to reproduce the bug again to post the log is that I was wrong, it is not caused just by using the hotkey to change from fullscreen to window mode and back. It happens when using it in combination with the toggle capture hotkey.

So to reproduce the bug follow this:

1 - Open new fullscreen stream session with special windows hotkeys activated
2 - Ctrl+Alt+Shift+x to change fullscreen to window mode
3 - Ctrl+Alt+Shift+z to release input capture from host session
4 - Click within host window session to capture again input on host
5 - Ctrl+Alt+Shift+x to change back from window mode to fullscreen
6 - Press Win key and it will work as expected on host
7 - Now press just any cursor key will make the screen completely black but actually it will start changing screen rotation on client PC
8 - Ctrl+Alt+Shift+q will exit stream session leaving the screen orientation from the last cursor key pressed before closing session.

Also adding now log and screenshot of options screen:

[Moonlight-1614675126.log](https://github.com/moonlight-stream/moonlight-qt/files/6067484/Moonlight-1614675126.log)
![moonlight_options](https://user-images.githubusercontent.com/65514542/109625902-2b544b00-7b40-11eb-8037-624f52c2d5ff.PNG)

Hope to be more useful this time regarding this issue.

**Comment by cgutman** [comment]
It's no big deal. I've actually rewritten the keyboard hooking code since v3.0.0 in the process of upstreaming it to SDL to make it more robust against cases like this (and add new features too).

Can you try your scenario with a recent nightly build like https://ci.appveyor.com/project/cgutman/moonlight-qt/builds/38032498/job/wcs06kw1yt23fq10/artifacts

If it still happens, grab a log with `set SDL_EVENT_LOGGING=1` like before.

**Comment by Jeslandor** [comment]
Ok, I finally was able to play around a bit with the build you linked and I have to say that as far as I tested the bug/problem seems to be gone. Tried both options (fullscreen and always) and everything worked as expected.

Thank you very much.

**Comment by cgutman** [comment]
Fix released in v3.1.0.
